provider/reprovider.go

```diff
@@ -236,208 +302,223 @@ func initialReprovideDelay(duration time.Duration) Option {
-					s.throughputCallback = nil
-				}
-				s.throughputProvideCurrentCount = 0
-				s.throughputDurationSum = 0
+			// Deduplicate recently provided cids
+			if found, _ := s.instantProvideDeduplicator.ContainsOrAdd(c, struct{}{}); found {
+				continue
 			}
+			provideOperation(s.ctx, c)
+		case <-s.ctx.Done():
+			return
 		}
-	}()
+	}
+}
 
-	// don't start reprovide scheduling if reprovides are disabled (reprovideInterval == 0)
-	if s.reprovideInterval == 0 {
+func (s *reprovider) reprovideSchedulingWorker() {
+	defer s.closewg.Done()
+
+	// read last reprovide time written to the datastore, and schedule the
+	// first reprovide to happen reprovideInterval after that
+	firstReprovideDelay := s.initalReprovideDelay
+	lastReprovide, err := s.getLastReprovideTime()
+	if err == nil && time.Since(lastReprovide) < s.reprovideInterval-s.initalReprovideDelay {
+		firstReprovideDelay = time.Until(lastReprovide.Add(s.reprovideInterval))
+	}
+	firstReprovideTimer := time.NewTimer(firstReprovideDelay)
+
+	select {
+	case <-firstReprovideTimer.C:
+	case <-s.ctx.Done():
 		return
 	}
 
-	s.closewg.Add(1)
-	go func() {
-		// reprovides scheduling worker
-		defer s.closewg.Done()
-
-		// read last reprovide time written to the datastore, and schedule the
-		// first reprovide to happen reprovideInterval after that
-		firstReprovideDelay := s.initalReprovideDelay
-		lastReprovide, err := s.getLastReprovideTime()
-		if err == nil && time.Since(lastReprovide) < s.reprovideInterval-s.initalReprovideDelay {
-			firstReprovideDelay = time.Until(lastReprovide.Add(s.reprovideInterval))
-		}
-		firstReprovideTimer := time.NewTimer(firstReprovideDelay)
+	// after the first reprovide, schedule periodical reprovides
+	nextReprovideTicker := time.NewTicker(s.reprovideInterval)
 
+	for {
+		err := s.Reprovide(context.Background())
+		if err != nil {
+			if s.ctx.Err() != nil {
+				return
+			}
+			log.Errorf("failed to reprovide: %s", err)
+		}
 		select {
-		case <-firstReprovideTimer.C:
+		case <-nextReprovideTicker.C:
 		case <-s.ctx.Done():
 			return
 		}
+	}
+}
 
-		// after the first reprovide, schedule periodical reprovides
-		nextReprovideTicker := time.NewTicker(s.reprovideInterval)
+func (s *reprovider) run() {
+	s.closewg.Add(1)
+	if s.batchProvides {
+		go s.batchProvideWorker()
+	} else {
+		go s.instantProvideWorker()
+	}
 
-		for {
-			err := s.Reprovide(context.Background())
-			if err != nil {
-				if s.ctx.Err() != nil {
-					return
-				}
-				log.Errorf("failed to reprovide: %s", err)
-			}
+	// don't start reprovide scheduling if reprovides are disabled (reprovideInterval == 0)
+	if s.reprovideInterval > 0 {
+		s.closewg.Add(1)
+		go s.reprovideSchedulingWorker()
+	}
+}
+
+func (s *reprovider) waitUntilProvideSystemReady() {
+	if r, ok := s.rsys.(Ready); ok {
+		ticker := time.NewTicker(time.Minute)
+		for !r.Ready() {
```

Comment:
Consider not creating the ticker until after the first ready check fails.
```suggestion
		var ticker time.Ticker
		for !r.Ready() {
			if ticker == nil {
				ticker = time.NewTicker(time.Minute)
				defer ticker.Stop()
			}
```